### DISCUSSION TITLE
Multi-pairing with provable pairing

### DISCUSSION THREAD

**Discussion by shramee** [Consensys/gnark#1347] [General]
Hi guys,

I think [`millerLoopAndFinalExpResult`](https://github.com/Consensys/gnark/blob/754b26e0c06785c1d5d1364228aa75fec01dac7c/std/algebra/emulated/sw_bn254/pairing.go#L686) should do the most optimised circuit for (emulated?) Pairings. 
I was wondering why this one doesn't support multiple pairings..?

i.e. Instead of,
```go
func (pr Pairing) IsMillerLoopAndFinalExpOne(P *G1Affine, Q *G2Affine, previous *GTEl) frontend.Variable {
```
why don't we have P and Q as arrays in in the main pairing func?
```go
func (pr Pairing) IsMillerLoopAndFinalExpOne(P []*G1Affine, Q []*G2Affine, previous *GTEl) frontend.Variable {
```

@yelhousni @ivokub

**Comment by ivokub** [comment]
Internally, in case of multi pairing check the algorithm is:
```
accumulator := Ext12.One()
for _, pairs := inputs {
    millerLoopResult := pairing.MillerLoop(pairs.P, pairs.Q)
    accumulator = Ext12.Mul(millerLoopResult, accumulator)
}
toTargetGroup := finalExp(accumulator)
Ext12.AssertIsOne(toTargetGroup)
```

the method `IsMillerLoopAndFinalExpOne` combines the last Miller loop computation, accumulator multiplication and final exp check as a single method. It is very specific to Linea use case and normally not needed elsewhere. There we want to have flexibility in the number of pairs we want to allow in pairing check and then we need to perform miller loops and finalexps separately.

For generic multi pairing check you should use [`PairingCheck`](https://github.com/Consensys/gnark/blob/master/std/algebra/interfaces.go#L98) method which is the full algorithm.

**Reply by ivokub** [reply]
If you want to have a look how we do it in Linea, then you can check here https://github.com/Consensys/linea-monorepo/tree/main/prover/zkevm/prover/ecpair. We currently don't have published the full documentation which describes it further, but it will be probably soon.

**Reply by shramee** [reply]
Aah thanks!
I hadn't seen that. Though it seems to do additional ~64 FQ12 squarings than necessary but looks good.

[Exp by 6u+2](https://github.com/Consensys/gnark/blob/f3d91999250b51d7457d3b945469e7a0d567ef38/std/algebra/emulated/sw_bn254/pairing.go#L201) could be in the miller loop too like in [millerLoopAndFinalExpResult](https://github.com/Consensys/gnark/blob/754b26e0c06785c1d5d1364228aa75fec01dac7c/std/algebra/emulated/sw_bn254/pairing.go#L686)?

**Reply by ivokub** [reply]
Hmm, indeed may be. @yelhousni?

**Reply by yelhousni** [reply]
So there are two things:
1- In linea we need a fixed circuit for ECPair precompile regardless of the number of inputs. We cannot have this if we compute a multi-Miller loop. So the [solution](
https://github.com/consensys/gnark/tree/3729f222cdaeb4debfb58ce4f3ed380a0da716cb/std/evmprecompiles/08-bnpairing.go#L12) we ended up with is to have 2 fixed circuits. `MillerLoopAndMul` computes the Miller loop of 2 inputs and multiplies it by the previous ML result (1 for the first call). `MillerLoopAndFinalExpCheck` computes the Miller loop of 2 inputs and checks that the final exp is 1 using Novakovic-Eagen paper (+ including the squarings in the Miller loop).
2- Independently of Linea we can have a `PairingCheck` method that is similar to `MillerLoopAndFinalExpCheck` but for slices of points. The problem is you need to have all the (arbitrary number) of inputs in the hint. It's just an Go implementation detail that I haven't figured out yet. If you have an idea you're more that welcome to push a PR! I did a [PR](https://github.com/Consensys/gnark/pull/1230) for `DoublePairingCheck` with just 4 inputs since it's the main use of pairings in protocols but it's still uses the old pairing Fpk tower.

**Reply by yelhousni** [reply]
Here you go for 2-
https://github.com/Consensys/gnark/pull/1365

**Reply by shramee** [reply]
Sweet! The PR looks real nice!

**Reply by yelhousni** [reply]
@shramee I can't put you as a reviewer on the PR apparently but feel free to add your review if you want!

**Reply by shramee** [reply]
Thanks @yelhousni :D I've added it!
